### PR TITLE
#93 Add new meeting page

### PR DESCRIPTION
**Pull request by olena-kozak**
image
image
image

### CODE REVIEW

File: frontend/src/app/core/helpers/date-helper.ts [hunk 1]
```diff
@@ -0,0 +1,2 @@
+export const getDisplayDate = (): string[] =>
+    ['24 Aug 2022', '25 Aug 2022', '26 Aug 2022', '27 Aug 2022', '28 Aug 2022', '29 Aug 2022', '30 Aug 2022'];
```
Comment (tatianahutii) [inline frontend/src/app/core/helpers/date-helper.ts]: Do we still use this?

File: frontend/src/app/modules/new-meeting/new-meeting/new-meeting.component.ts [hunk 2]
```diff
@@ -0,0 +1,207 @@
...
+
+@Component({
+    selector: 'app-new-meeting',
+    templateUrl: './new-meeting.component.html',
+    styleUrls: ['./new-meeting.component.sass'],
+})
+export class NewMeetingComponent extends BaseComponent implements OnInit {
+    constructor(private newMeetingService: NewMeetingService, public notificationService: NotificationService) {
+        super();
+        this.getTeamMembersOfCurrentUser();
```
Comment (tatianahutii) [inline frontend/src/app/modules/new-meeting/new-meeting/new-meeting.component.ts]: Why in the constructor?

File: frontend/src/app/modules/new-meeting/new-meeting/new-meeting.component.ts [hunk 3]
```diff
@@ -0,0 +1,207 @@
...
+
+    public filteredOptions: Observable<INewMeetingMember[]>;
+
+    public durations: IDuration[] = getDisplayDuration();
+
+    public dates = getDisplayDate();
+
+    public locations = Object.keys(LocationType).filter(key => Number.isNaN(Number(key)));
+
+    public unitOfTime = Object.keys(UnitOfTime).filter(key => Number.isNaN(Number(key)));
```
Comment (tatianahutii) [inline frontend/src/app/modules/new-meeting/new-meeting/new-meeting.component.ts]: Why do we need a filter here?

File: frontend/src/app/modules/new-meeting/new-meeting/new-meeting.component.ts [hunk 4]
```diff
@@ -0,0 +1,207 @@
...
+                this.teamMembers = resp;
+
+                this.filteredOptions = this.memberFilterCtrl.valueChanges.pipe(
+                    startWith(''),
+                    map(value => {
+                        const name = typeof value === 'string' ? value : value?.name;
+
+                        return name ? this._filter(name as string) : this.teamMembers.slice();
+                    }),
+                );
```
Comment (tatianahutii) [inline frontend/src/app/modules/new-meeting/new-meeting/new-meeting.component.ts]: Move this to a separate method
This logic looks a little bit strange try to refactor this

File: frontend/src/app/modules/new-meeting/new-meeting/new-meeting.component.ts [hunk 5]
```diff
@@ -0,0 +1,207 @@
...
+        } else {
+            this.customTimeShown = false;
+            this.durationChanged(durationValue.time, durationValue.unitOfTime);
+        }
+
+        if (mainContainerDurationValue.time === 'Custom') {
+            this.mainContentCustomTimeShown = true;
+        } else {
+            this.mainContentCustomTimeShown = false;
+        }
```
Comment (tatianahutii) [inline frontend/src/app/modules/new-meeting/new-meeting/new-meeting.component.ts]:
```suggestion
        this.mainContentCustomTimeShown = mainContainerDurationValue.time === 'Custom'
```

File: frontend/yarn.lock [hunk 6]
```diff
@@ -6070,6 +6070,13 @@ ngx-mask@^14.0.3:
   dependencies:
     tslib "^2.3.0"
 
+ngx-mat-select-search@^4.2.1:
```
Comment (tatianahutii) [inline frontend/yarn.lock]: Why did you add a new package?

**Review by tatianahutii — CHANGES_REQUESTED** [review]

File: frontend/src/app/core/helpers/date-helper.ts (hunk 1 above)
Comment (olena-kozak) [reply frontend/src/app/core/helpers/date-helper.ts]: Already not. Removed it.

File: frontend/src/app/modules/new-meeting/new-meeting/new-meeting.component.ts (hunk 5 above)
Comment (olena-kozak) [reply frontend/src/app/modules/new-meeting/new-meeting/new-meeting.component.ts]: Okay, refactored it.

File: frontend/yarn.lock (hunk 6 above)
Comment (olena-kozak) [reply frontend/yarn.lock]: Already not. Removed it.

File: frontend/src/app/modules/new-meeting/new-meeting/new-meeting.component.ts (hunk 3 above)
Comment (olena-kozak) [reply frontend/src/app/modules/new-meeting/new-meeting/new-meeting.component.ts]: I need a filter to display enums properly in element
Before using filter:
image
After using filter:
image

File: frontend/src/app/modules/new-meeting/new-meeting/new-meeting.component.ts (hunk 4 above)
Comment (olena-kozak) [reply frontend/src/app/modules/new-meeting/new-meeting/new-meeting.component.ts]: Done.

File: frontend/src/app/modules/new-meeting/new-meeting/new-meeting.component.ts (hunk 2 above)
Comment (olena-kozak) [reply frontend/src/app/modules/new-meeting/new-meeting/new-meeting.component.ts]: Removed calling this method from constructor.

File: frontend/src/app/modules/new-meeting/new-meeting/new-meeting.component.ts (hunk 3 above)
Comment (tatianahutii) [reply frontend/src/app/modules/new-meeting/new-meeting/new-meeting.component.ts]: This is the problem with using enums
We don't need filter here

File: frontend/src/app/modules/new-meeting/new-meeting/new-meeting.component.ts [hunk 7]
```diff
@@ -0,0 +1,196 @@
...
+        this.newMeetingService
+            .getTeamMembersOfCurrentUser()
+            .pipe(this.untilThis)
+            .subscribe((resp) => {
+                this.teamMembers = resp;
+                this._getFilteredOptions();
+            });
+    }
+
+    private _getFilteredOptions() {
```
Comment (tatianahutii) [inline frontend/src/app/modules/new-meeting/new-meeting/new-meeting.component.ts]:
```suggestion
    private getFilteredOptions() {
```

Comment (tatianahutii) [reply frontend/src/app/modules/new-meeting/new-meeting/new-meeting.component.ts]: Private methods should go after public

File: frontend/src/app/modules/new-meeting/new-meeting/new-meeting.component.ts [hunk 8]
```diff
@@ -0,0 +1,196 @@
...
+            map(value => {
+                const filterValue = value || ''.toLowerCase();
+
+                return this.teamMembers.filter(teamMembers => teamMembers.name.toLowerCase().includes(filterValue));
+            }),
+        );
+    }
+
+    public displayMemberName(teamMember: INewMeetingMember): string {
+        return teamMember && teamMember.name ? teamMember.name : '';
```
Comment (tatianahutii) [inline frontend/src/app/modules/new-meeting/new-meeting/new-meeting.component.ts]:
```suggestion
        return teamMember? .name ??  '';
```

File: frontend/src/app/modules/new-meeting/new-meeting/new-meeting.component.ts [hunk 9]
```diff
@@ -0,0 +1,196 @@
...
+
+        this.mainContentCustomTimeShown = form.value.mainContainerDuration.time === 'Custom';
+
+        if (durationValue.time === 'Custom') {
+            this.customTimeShown = true;
+            this.setValidation();
+        } else {
+            this.customTimeShown = false;
+            this.durationChanged(durationValue.time, durationValue.unitOfTime);
+        }
```
Comment (tatianahutii) [inline frontend/src/app/modules/new-meeting/new-meeting/new-meeting.component.ts]:
```suggestion
        this.customTimeShown = durationValue.time === 'Custom';
        if (this.customTimeShown) {
            this.setValidation();
        } else {
            this.durationChanged(durationValue.time, durationValue.unitOfTime);
        }
```

**Review by tatianahutii — CHANGES_REQUESTED** [review]

**Review by apofis96 — APPROVED** [review]

File: frontend/src/app/modules/new-meeting/new-meeting/new-meeting.component.ts (hunk 7 above)
Comment (olena-kozak) [reply frontend/src/app/modules/new-meeting/new-meeting/new-meeting.component.ts]: Done.

File: frontend/src/app/modules/new-meeting/new-meeting/new-meeting.component.ts (hunk 8 above)
Comment (olena-kozak) [reply frontend/src/app/modules/new-meeting/new-meeting/new-meeting.component.ts]: Refactored it.

File: frontend/src/app/modules/new-meeting/new-meeting/new-meeting.component.ts (hunk 9 above)
Comment (olena-kozak) [reply frontend/src/app/modules/new-meeting/new-meeting/new-meeting.component.ts]: Refactored it.

File: frontend/src/app/modules/new-meeting/new-meeting/new-meeting.component.ts (hunk 3 above)
Comment (olena-kozak) [reply frontend/src/app/modules/new-meeting/new-meeting/new-meeting.component.ts]: Refactored it.

File: frontend/src/app/modules/new-meeting/new-meeting/new-meeting.component.ts [hunk 10]
```diff
@@ -30,9 +30,12 @@ export class NewMeetingComponent extends BaseComponent implements OnInit {
...
     public durations: IDuration[] = getDisplayDuration();
 
-    public locations = Object.keys(LocationType).filter(key => Number.isNaN(Number(key)));
+    public locations: string[] = [
+        LocationType[LocationType.Zoom],
+        LocationType[LocationType.GoogleMeet],
+        LocationType[LocationType.Office]];
 
-    public unitOfTime = Object.keys(UnitOfTime).filter(key => Number.isNaN(Number(key)));
+    public unitOfTime: string[] = [UnitOfTime[UnitOfTime.min], UnitOfTime[UnitOfTime.hour]];
```
Comment (tatianahutii) [inline frontend/src/app/modules/new-meeting/new-meeting/new-meeting.component.ts]: it would be better to change  enums or use something like keyof typeof {{enum name}}

**Review by tatianahutii — CHANGES_REQUESTED** [review]

File: frontend/src/app/modules/new-meeting/new-meeting/new-meeting.component.ts (hunk 10 above)
Comment (olena-kozak) [reply frontend/src/app/modules/new-meeting/new-meeting/new-meeting.component.ts]: But I don't have this code in file now. I refactored it to using labelMapper.

File: frontend/src/app/modules/new-meeting/new-meeting/new-meeting.component.ts [hunk 11]
```diff
@@ -0,0 +1,195 @@
...
+    public removeMemberToList(form: FormGroup) {
+        this.addedMembers = this.addedMembers.filter((member) => member.id !== form.value.teamMember.id);
+    }
+
+    public reset() {
+        this.meetingForm.reset();
+        this.patchFormValues();
+        this.addedMembers = [];
+    }
+
```
Comment (va-dym-ko-val) [inline frontend/src/app/modules/new-meeting/new-meeting/new-meeting.component.ts:184]: You can omit "public" for public methods

**Review by va-dym-ko-val — CHANGES_REQUESTED** [review]

**Review by va-dym-ko-val — APPROVED** [review]

**Review by tatianahutii — APPROVED** [review]
